File: backend/app/schemas/knowledge.py
"""Pydantic v2 schemas for the Knowledge Base API."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, HttpUrl, field_validator, model_validator


# ── Shared ────────────────────────────────────────────────────────────────────
ArticleStatus     = Literal["draft", "review", "approved", "published", "archived"]
ArticleVisibility = Literal["public", "internal", "agent_only", "customer_specific"]
LinkType          = Literal["related", "resolved_by", "referenced"]


# ── Category ──────────────────────────────────────────────────────────────────
class CategoryTranslationIn(BaseModel):
    language: str = Field(..., min_length=2, max_length=10)
    name: str     = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None


class CategoryTranslationOut(CategoryTranslationIn):
    id: UUID
    model_config = ConfigDict(from_attributes=True)


class CategoryCreate(BaseModel):
    parent_id: Optional[UUID] = None
    slug: str                 = Field(..., min_length=1, max_length=255, pattern=r"^[a-z0-9-]+$")
    icon: Optional[str]       = Field(None, max_length=100)
    sort_order: int           = 0
    translations: list[CategoryTranslationIn] = Field(default_factory=list)


class CategoryUpdate(BaseModel):
    parent_id: Optional[UUID]  = None
    slug: Optional[str]        = Field(None, pattern=r"^[a-z0-9-]+$")
    icon: Optional[str]        = None
    sort_order: Optional[int]  = None
    translations: Optional[list[CategoryTranslationIn]] = None


class CategoryOut(BaseModel):
    id: UUID
    parent_id: Optional[UUID]
    slug: str
    icon: Optional[str]
    sort_order: int
    translations: list[CategoryTranslationOut]
    children: list["CategoryOut"] = []
    article_count: int = 0
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


# ── Translation ───────────────────────────────────────────────────────────────
class TranslationCreate(BaseModel):
    language: str   = Field(..., min_length=2, max_length=10)
    title: str      = Field(..., min_length=1, max_length=500)
    content: str    = Field(..., min_length=1)
    excerpt: Optional[str] = Field(None, max_length=500)


class TranslationUpdate(BaseModel):
    title: Optional[str]   = Field(None, min_length=1, max_length=500)
    content: Optional[str] = Field(None, min_length=1)
    excerpt: Optional[str] = None


class TranslationOut(BaseModel):
    id: UUID
    language: str
    title: str
    content: str
    excerpt: Optional[str]
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


# ── Attachment ────────────────────────────────────────────────────────────────
class AttachmentOut(BaseModel):
    id: UUID
    file_name: str
    file_path: str
    file_size: Optional[int]
    content_type: Optional[str]
    is_image: bool
    uploaded_by: UUID
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


# ── Tag ───────────────────────────────────────────────────────────────────────
class TagCreate(BaseModel):
    name: str             = Field(..., min_length=1, max_length=100)
    color: Optional[str]  = Field(None, max_length=20)


class TagOut(BaseModel):
    id: UUID
    name: str
    color: Optional[str]
    model_config = ConfigDict(from_attributes=True)


# ── Version ───────────────────────────────────────────────────────────────────
class VersionOut(BaseModel):
    id: UUID
    version_number: int
    language: str
    title: str
    content: str
    change_summary: Optional[str]
    changed_by: UUID
    changed_by_name: Optional[str] = None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class RevertRequest(BaseModel):
    change_summary: Optional[str] = "Reverted to previous version"


# ── Article ───────────────────────────────────────────────────────────────────
class ArticleCreate(BaseModel):
    category_id: Optional[UUID]     = None
    slug: str                        = Field(..., min_length=1, max_length=600, pattern=r"^[a-z0-9-]+$")
    status: ArticleStatus            = "draft"
    visibility: ArticleVisibility    = "public"
    default_language: str            = "en"
    reference_url: Optional[str]     = None
    sort_order: int                  = 0
    tag_ids: list[UUID]              = Field(default_factory=list)
    translations: list[TranslationCreate] = Field(..., min_length=1)
    change_summary: Optional[str]    = None

    @field_validator("reference_url")
    @classmethod
    def validate_url(cls, v: str | None) -> str | None:
        if v and not v.startswith(("http://", "https://")):
            raise ValueError("reference_url must be a valid HTTP(S) URL")
        return v


class ArticleUpdate(BaseModel):
    category_id: Optional[UUID]          = None
    slug: Optional[str]                  = Field(None, pattern=r"^[a-z0-9-]+$")
    status: Optional[ArticleStatus]      = None
    visibility: Optional[ArticleVisibility] = None
    default_language: Optional[str]      = None
    reference_url: Optional[str]         = None
    sort_order: Optional[int]            = None
    tag_ids: Optional[list[UUID]]        = None
    translations: Optional[list[TranslationCreate]] = None
    change_summary: Optional[str]        = None


class ArticleAuthorOut(BaseModel):
    id: UUID
    name: str
    model_config = ConfigDict(from_attributes=True)


class ArticleOut(BaseModel):
    id: UUID
    category_id: Optional[UUID]
    category: Optional[CategoryOut]
    author_id: UUID
    author: Optional[ArticleAuthorOut]
    slug: str
    status: str
    visibility: str
    default_language: str
    reference_url: Optional[str]
    sort_order: int
    view_count: int
    helpful_yes: int
    helpful_no: int
    published_at: Optional[datetime]
    reviewed_by: Optional[UUID]
    reviewed_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime
    translations: list[TranslationOut]  = []
    tags: list[TagOut]                   = []
    attachments: list[AttachmentOut]     = []
    version_count: int                   = 0
    model_config = ConfigDict(from_attributes=True)


class ArticleListOut(BaseModel):
    """Lightweight list view — no content body."""
    id: UUID
    category_id: Optional[UUID]
    slug: str
    status: str
    visibility: str
    default_language: str
    view_count: int
    helpful_yes: int
    helpful_no: int
    published_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime
    author: Optional[ArticleAuthorOut]
    tags: list[TagOut]             = []
    # Active-language translation header (title + excerpt)
    title: str                     = ""
    excerpt: Optional[str]         = None
    model_config = ConfigDict(from_attributes=True)


class PaginatedArticles(BaseModel):
    items: list[ArticleListOut]
    total: int
    page: int
    page_size: int
    pages: int


# ── Search ────────────────────────────────────────────────────────────────────
class SearchHit(BaseModel):
    article_id: UUID
    slug: str
    language: str
    title: str
    headline: str          # ts_headline snippet
    rank: float
    status: str
    visibility: str
    category_id: Optional[UUID]
    category_name: Optional[str]
    published_at: Optional[datetime]


class SearchResult(BaseModel):
    query: str
    hits: list[SearchHit]
    total: int
    took_ms: float


# ── Ticket links ──────────────────────────────────────────────────────────────
class TicketLinkCreate(BaseModel):
    ticket_id: UUID
    link_type: LinkType = "related"


class TicketLinkOut(BaseModel):
    id: UUID
    ticket_id: UUID
    article_id: UUID
    link_type: str
    linked_by: UUID
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


# ── Feedback ──────────────────────────────────────────────────────────────────
class FeedbackCreate(BaseModel):
    is_helpful: bool
    comment: Optional[str] = Field(None, max_length=1000)


class FeedbackOut(BaseModel):
    id: UUID
    is_helpful: bool
    comment: Optional[str]
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


# ── Status transition ─────────────────────────────────────────────────────────
class StatusTransitionRequest(BaseModel):
    status: ArticleStatus
    comment: Optional[str] = None


# Allow forward references
CategoryOut.model_rebuild()

File: src/types/knowledge.ts
// ─── Shared literals ──────────────────────────────────────────────────────────
export type ArticleStatus     = 'draft' | 'review' | 'approved' | 'published' | 'archived'
export type ArticleVisibility = 'public' | 'internal' | 'agent_only' | 'customer_specific'
export type LinkType          = 'related' | 'resolved_by' | 'referenced'

// ─── Category ─────────────────────────────────────────────────────────────────
export interface CategoryTranslation {
  id: string
  language: string
  name: string
  description?: string | null
}

export interface Category {
  id: string
  parent_id: string | null
  slug: string
  icon?: string | null
  sort_order: number
  translations: CategoryTranslation[]
  children: Category[]
  article_count: number
  created_at: string
  updated_at: string
}

export interface CategoryCreate {
  parent_id?: string | null
  slug: string
  icon?: string
  sort_order?: number
  translations: { language: string; name: string; description?: string }[]
}

export interface CategoryUpdate extends Partial<CategoryCreate> {}

// ─── Translation ──────────────────────────────────────────────────────────────
export interface ArticleTranslation {
  id: string
  language: string
  title: string
  content: string
  excerpt?: string | null
  created_at: string
  updated_at: string
}

export interface TranslationInput {
  language: string
  title: string
  content: string
  excerpt?: string
}

// ─── Attachment ───────────────────────────────────────────────────────────────
export interface Attachment {
  id: string
  file_name: string
  file_path: string
  file_size?: number
  content_type?: string
  is_image: boolean
  uploaded_by: string
  created_at: string
}

// ─── Tag ──────────────────────────────────────────────────────────────────────
export interface Tag {
  id: string
  name: string
  color?: string | null
}

// ─── Version ──────────────────────────────────────────────────────────────────
export interface ArticleVersion {
  id: string
  version_number: number
  language: string
  title: string
  content: string
  change_summary?: string | null
  changed_by: string
  changed_by_name?: string | null
  created_at: string
}

// ─── Author ───────────────────────────────────────────────────────────────────
export interface ArticleAuthor {
  id: string
  name: string
}

// ─── Article (full) ───────────────────────────────────────────────────────────
export interface Article {
  id: string
  category_id?: string | null
  category?: Category | null
  author_id: string
  author?: ArticleAuthor | null
  slug: string
  status: ArticleStatus
  visibility: ArticleVisibility
  default_language: string
  reference_url?: string | null
  sort_order: number
  view_count: number
  helpful_yes: number
  helpful_no: number
  published_at?: string | null
  reviewed_by?: string | null
  reviewed_at?: string | null
  created_at: string
  updated_at: string
  translations: ArticleTranslation[]
  tags: Tag[]
  attachments: Attachment[]
  version_count: number
}

// ─── Article (list) ───────────────────────────────────────────────────────────
export interface ArticleListItem {
  id: string
  category_id?: string | null
  slug: string
  status: ArticleStatus
  visibility: ArticleVisibility
  default_language: string
  view_count: number
  helpful_yes: number
  helpful_no: number
  published_at?: string | null
  created_at: string
  updated_at: string
  author?: ArticleAuthor | null
  tags: Tag[]
  title: string
  excerpt?: string | null
}

export interface PaginatedArticles {
  items: ArticleListItem[]
  total: number
  page: number
  page_size: number
  pages: number
}

// ─── Article create / update ──────────────────────────────────────────────────
export interface ArticleCreate {
  category_id?: string | null
  slug: string
  status?: ArticleStatus
  visibility?: ArticleVisibility
  default_language?: string
  reference_url?: string
  sort_order?: number
  tag_ids?: string[]
  translations: TranslationInput[]
  change_summary?: string
}

export interface ArticleUpdate extends Partial<ArticleCreate> {}

// ─── Search ───────────────────────────────────────────────────────────────────
export interface SearchHit {
  article_id: string
  slug: string
  language: string
  title: string
  headline: string            // HTML with <mark> highlights
  rank: number
  status: ArticleStatus
  visibility: ArticleVisibility
  category_id?: string | null
  category_name?: string | null
  published_at?: string | null
}

export interface SearchResult {
  query: string
  hits: SearchHit[]
  total: number
  took_ms: number
}

// ─── Ticket link ──────────────────────────────────────────────────────────────
export interface TicketLink {
  id: string
  ticket_id: string
  article_id: string
  link_type: LinkType
  linked_by: string
  created_at: string
}

export interface TicketArticleLink {
  link_id: string
  link_type: LinkType
  article_id: string
  slug: string
  status: ArticleStatus
  visibility: ArticleVisibility
  title: string
  linked_at: string
}

// ─── Feedback ─────────────────────────────────────────────────────────────────
export interface FeedbackCreate {
  is_helpful: boolean
  comment?: string
}

// ─── List filters ─────────────────────────────────────────────────────────────
export interface ArticleFilters {
  category_id?: string
  status?: ArticleStatus | ''
  tag_id?: string
  language?: string
  page?: number
  page_size?: number
}

export interface SearchParams {
  q: string
  language?: string
  category_id?: string
  status?: string
  limit?: number
  offset?: number
}

// ─── Status / visibility meta ─────────────────────────────────────────────────
export const STATUS_META: Record<ArticleStatus, { label: string; cls: string }> = {
  draft:     { label: 'Draft',     cls: 'bg-slate-400/15 text-slate-400 border border-slate-400/30' },
  review:    { label: 'Review',    cls: 'bg-amber-400/15 text-amber-400 border border-amber-400/30' },
  approved:  { label: 'Approved',  cls: 'bg-blue-400/15 text-blue-400 border border-blue-400/30' },
  published: { label: 'Published', cls: 'bg-emerald-400/15 text-emerald-400 border border-emerald-400/30' },
  archived:  { label: 'Archived',  cls: 'bg-rose-400/15 text-rose-400 border border-rose-400/30' },
}

export const VISIBILITY_META: Record<ArticleVisibility, { label: string; cls: string }> = {
  public:            { label: 'Public',            cls: 'bg-indigo-400/15 text-indigo-400 border border-indigo-400/30' },
  internal:          { label: 'Internal',          cls: 'bg-amber-400/15 text-amber-400 border border-amber-400/30' },
  agent_only:        { label: 'Agent Only',        cls: 'bg-violet-400/15 text-violet-400 border border-violet-400/30' },
  customer_specific: { label: 'Customer Specific', cls: 'bg-teal-400/15 text-teal-400 border border-teal-400/30' },
}
